price-feeder/oracle/provider/ftx.go

```diff
@@ -0,0 +1,344 @@
+	if err != nil {
+		return time.Time{}, fmt.Errorf("unable to parse ftx timestamp")
+	}
+	return t, nil
+}
+
+func NewFTXProvider(
+	ctx context.Context,
+	logger zerolog.Logger,
+	endpoint Endpoint,
+	pairs ...types.CurrencyPair,
+) *FTXProvider {
+	restURL := ftxRestURL
+
+	if endpoint.Name == ProviderFTX {
+		restURL = endpoint.Rest
+	}
+
+	ftx := FTXProvider{
+		baseURL:      restURL,
+		client:       newDefaultHTTPClient(),
+		logger:       logger,
+		candleCache:  nil,
+		marketsCache: []FTXMarkets{},
+	}
+
+	go func() {
+		logger.Debug().Msg("starting ftx polling...")
+		err := ftx.pollCache(ctx, pairs...)
+		if err != nil {
+			logger.Err(err).Msg("ftx provider unable to poll new data")
+		}
+	}()
+
+	return &ftx
+}
+
+// GetCandlePrices returns the tickerPrices based on the provided pairs.
+func (p *FTXProvider) GetTickerPrices(pairs ...types.CurrencyPair) (map[string]types.TickerPrice, error) {
+	markets := p.getMarketsCache()
+
+	baseDenomIdx := make(map[string]types.CurrencyPair)
+	for _, cp := range pairs {
+		baseDenomIdx[strings.ToUpper(cp.Base)] = cp
+	}
+
+	tickerPrices := make(map[string]types.TickerPrice, len(pairs))
+	for _, tr := range markets {
+		symbol := strings.ToUpper(tr.Base)
+
+		cp, ok := baseDenomIdx[symbol]
+		if !ok {
+			// skip tokens that are not requested
+			continue
+		}
+
+		if _, ok := tickerPrices[symbol]; ok {
+			return nil, fmt.Errorf("duplicate token found in FTX response: %s", symbol)
+		}
+
+		priceRaw := strconv.FormatFloat(tr.Price, 'f', -1, 64)
+		price, err := sdk.NewDecFromStr(priceRaw)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read FTX price (%f) for %s", tr.Price, symbol)
+		}
+
+		volumeRaw := strconv.FormatFloat(tr.Volume, 'f', -1, 64)
+		volume, err := sdk.NewDecFromStr(volumeRaw)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read FTX volume (%f) for %s", tr.Volume, symbol)
+		}
+
+		tickerPrices[cp.String()] = types.TickerPrice{Price: price, Volume: volume}
+	}
+
+	for _, cp := range pairs {
+		if _, ok := tickerPrices[cp.String()]; !ok {
+			return nil, fmt.Errorf(types.ErrMissingExchangeRate.Error(), cp.String())
+		}
+	}
+
+	return tickerPrices, nil
+}
+
+// GetCandlePrices returns the cached candlePrices based on provided pairs.
+func (p *FTXProvider) GetCandlePrices(pairs ...types.CurrencyPair) (map[string][]types.CandlePrice, error) {
+	candleCache := p.getCandleCache()
+	if len(candleCache) < 1 {
+		return nil, fmt.Errorf("candles have not been cached")
+	}
+
+	candlePrices := make(map[string][]types.CandlePrice, len(pairs))
+	for _, pair := range pairs {
+		if _, ok := candleCache[pair.String()]; !ok {
+			return nil, fmt.Errorf("missing candles for %s", pair.String())
+		}
+		candlePrices[pair.String()] = candleCache[pair.String()]
+	}
+
+	return candlePrices, nil
+}
+
+// GetAvailablePairs return all available pairs symbol to susbscribe.
+func (p *FTXProvider) GetAvailablePairs() (map[string]struct{}, error) {
+	markets := p.getMarketsCache()
```

Comment:
we should hit `https://ftx.com/api/markets` to get all the possible available pairs in FTX here

Reply:
The cache that's filled in by `pollMarkets` does use this endpoint, so it's effectively getting that data. I just cached it since both `GetTickerPrices` and `GetAvailablePairs` use the same data here